Add bitmask module to nddata for manipulating (DQ) bit fields and creating masks

mcara opened the pull request:
This PR adds a `bitmask` module to `nddata` that provides utilities for handling DQ data and creating boolean masks. The purpose of this PR is to make this code available to the wider community. Currently this code lives in three locations: [`stsci.tools`](https://github.com/spacetelescope/stsci.tools/blob/5f25467705f592a5ba1487ba6d9eb9deff9f33d6/lib/stsci/tools/bitmask.py) (original), [`ccdproc`](https://github.com/astropy/ccdproc/blob/503968ebc67373ae8e4e2ff4759bf76998ed987a/ccdproc/extern/bitfield.py) (ported there), and [`jwst` pipeline](https://github.com/spacetelescope/jwst/blob/92f9b6f36694d6db18e5037a6c54d43983d5e671/jwst/resample/bitmask.py) (also ported there). This is an attempt to reduce all these copies to a single common location available to most users.

mcara commented:
CC: @mwcraig @crawfordsm @jdavies-st

mwcraig commented:
:tada: I'll review this tonight!

pllim commented:
Note: bot and CI are absent due to GitHub webhook issues. Should wait for them to get back online before merging.

mcara commented:
@mwcraig Thank you!

mcara commented:
As a note, this PR is also a result of the suggestion from https://github.com/astropy/ccdproc/pull/611#issuecomment-362977219

mcara commented:
This also addresses the issue of `jwst` pipeline depending on `stsci.tools` - see https://github.com/spacetelescope/jwst/issues/1508#issuecomment-378664571 and the discussion around it.

mcara commented:
@astrofrog It seems that this `bitmask` module would be useful for #4741 but I am not sure about the details of exactly what functionality is desired in #4741.

mcara commented:
@astrofrog It seems that this `bitmask` module would be useful for #4741 but I am not sure about the details of exactly what functionality is desired in #4741.

At this moment, the main functionality of the `bitmask` module is to facilitate "interpretation" of bit flags (create a single integer bitmask from bit fields) and then convert a DQ array to a boolean mask. Obviously this can be used to select data based on a DQ data and what bits should be used or ignored in the DQ array when creating a boolean mask.

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+    bit_flags : int, str, list, None
+        An integer bitmask or flag, `None`, a string of comma- or
+        '+'-separated list of integer bit flags, or a Python list of integer
+        bit flags. If `bit_flags` is a `str` and if it is prepended with '~',
```
`bit_flags` needs double back-ticks around it like this (sphinx interprets single backticks as links, not literal text like github/markdown does):
```
``bit_flags``
```

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+        '+'-separated list of integer bit flags, or a Python list of integer
+        bit flags. If `bit_flags` is a `str` and if it is prepended with '~',
+        then the output bitmask will have its bits flipped (compared to simple
+        sum of input flags). For input `bit_flags` that is already a bitmask
```
double backticks

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+        then the output bitmask will have its bits flipped (compared to simple
+        sum of input flags). For input `bit_flags` that is already a bitmask
+        or a Python list of bit flags, bit-flipping can be controlled through
+        `flip_bits` parameter.
```
double backticks

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+    flip_bits : bool, None
+        Indicates whether or not to flip the bits of the returned bitmask
+        obtained from input bit flags. This parameter must be set to `None`
+        when input `bit_flags` is either `None` or a Python list of flags.
```
Double backticks on `bit_flags`

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+    -------
+    bitmask : int or None
+        Returns and integer bit mask formed from the input bit value
+        or `None` if input `bit_flags` parameter is `None` or an empty string.
```
Double backticks on `bit_flags`

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+    bitmask : int or None
+        Returns and integer bit mask formed from the input bit value
+        or `None` if input `bit_flags` parameter is `None` or an empty string.
+        If input string value was prepended with '~' (or `flip_bits` was
```
Double backticks on `flip_bits`

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+        ``ignore_flags``) in the input ``bitfield`` array. When
+        ``good_mask_value`` is non-zero or `True` then values in the output
+        boolean mask array corresponding to "good" bit fields in ``bitfield``
+        will be `True` (if ``dtype`` is `numpy.bool_`) or 1 (if ``dtype`` is
```
Looks like (when I tried building documentation locally) that sphinx isn't finding a link to `numpy.bool_`. Maybe either use double backticks or change to `numpy.bool`?

mcara replied:
@mwcraig  ``np.bool`` does not resolve either. However, it may be better to switch to `np.bool` due to impossibility if getting this right: `dtype : data-type (Default = numpy.bool\_)`. It seems like sphinx is having problems with trailing underscores (but not everywhere - just in a few settings).

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+        of numerical type) and values of corresponding to "bad" flags will be
+        `False` (or 0). When ``good_mask_value`` is zero or `False` then the
+        values in the output boolean mask array corresponding to "good" bit
+        fields in ``bitfield`` will be `False` (if ``dtype`` is `numpy.bool_`)
```
Same comment as above about `numpy.bool_`

mcara commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+
+
+def bitfield_to_boolean_mask(bitfield, ignore_flags=0, flip_bits=None,
+                             good_mask_value=True, dtype=np.bool_):
```
@mwcraig  This is a complicated topic and likely without a perfect solution. I am not a fan or user of masked arrays (many versions ago they were _extremely_ slow; don't know what is the situation now). For most of the my use cases, I need to select valid data from an image to compute some statistics on these data. I have a feeling that this is also the most likely scenario of usage by others as well: it is unlikely people will want to select "bad" data for any kind of computations (well, except for masked arrays). (Personally) I would prefer to keep current default but I do not have strong feelings on this subject especially if there is a larger consensus on this.

mcara replied:
@mwcraig At the same time I see `nddata` itself is defining `mask` in a way compatible with `numpy`'s masked arrays: file:///Users/mcara/repo/astropy/docs/_build/html/nddata/nddata.html#mask So, maybe it does make sense to have default changed for consistency. Still I wish it wasn't the case...

mcara replied:
@mwcraig At the same time I see `nddata` itself is defining `mask` in a way compatible with `numpy`'s masked arrays: http://docs.astropy.org/en/stable/nddata/nddata.html#mask So, maybe it does make sense to have default changed for consistency. Still I wish it wasn't the case...

mcara commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+def bitfield_to_boolean_mask(bitfield, ignore_flags=0, flip_bits=None,
+                             good_mask_value=True, dtype=np.bool_):
+    """
+    bitfield_to_boolean_mask(bitfield, ignore_flags=None, flip_bits=None, \
```
@mwcraig If I remove it, then `dtype` is shown as `dtype=<class 'numpy.bool_'>` instead of simply `dtype=numpy.bool_`. Which is more desirable?

mwcraig replied:
Huh, interesting. In that case, keep it in I guess.

mcara commented:
@mwcraig wrote:

> - Do you have time to add some more tests that the conversion to a mask produces the correct results? The test here are very extensive at checking for proper handling of bad bitfield formats. That is definitely important, and the doctests check a bit of the functionality but a couple of test cases would be nice.
> - Do you have time to write an narrative documentation?

With regard to narrative documentation - I can try tonight but I am not a prolific writer and so I hope to produce a base level that you may improve later. By the way, where should this "narrative documentation" go? `rst`? at the top of the module?

With regard to tests, you say you would like _"... more tests that the conversion to a mask produces the correct results"_. This is where I do not know what you feel is missing or not tested? The most critical part is actually deriving that integer "bitfield" value. The conversion to boolean mask essentially is happening in `bitwise_and()` here: https://github.com/astropy/astropy/blob/68a29760dc8c2b1a313d2cc34bcf19443892ee10/astropy/nddata/bitmask.py#L406 and the most important part is to figure out `ignore_mask`.

I simply do not see how additional tests could provide a meaningful improvement to what is already tested. However, it is quite possible that I am missing something, especially some edge cases. If so, I am open to specific suggestions to _meaningfully_ improve the tests.

mcara commented:
@mwcraig In the latest commit, I added `bitmask.rst` (no narrative documentation yet), put parameter names and `numpy` types into double back-quotes, removed `six` (no more Python 2.7 compatibility) and changed the default "good pixel" value to `False`.

mwcraig commented:
> With regard to narrative documentation - I can try tonight but I am not a prolific writer and so I hope to produce a base level that you may improve later. By the way, where should this "narrative documentation" go? rst? at the top of the module?

In the bitmask.rst file is fine for now. I can iterate on whatever you draft. Just including a couple of examples showing the different ways one can specify `ignore_flags` would be helpful. I'll take care of getting it linked into the rest of nddata documentation.

> I simply do not see how additional tests could provide a meaningful improvement to what is already tested. However, it is quite possible that I am missing something, especially some edge cases. If so, I am open to specific suggestions to meaningfully improve the tests.

I have to confess I didn't think that comment through carefully enough. Trying it out a bit this evening I haven't found any edge cases (I have one suggestion for adding a little more information in an error message I'll make as a line comment).

I'm checking on how to handle the failing tests (which are unrelated to your PR).

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+    bitmask = 0
+    for v in bitset:
+        if not is_bit_flag(v) and not allow_non_flags:
+            raise ValueError("Input list contains invalid (not powers of two) "
```
Could you please include the value that was invalid in the error message?

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+    Returns
+    -------
+    bitmask : int or None
+        Returns and integer bit mask formed from the input bit value or `None`
```
and → an

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+))
+
+
+def is_bit_flag(n):
```
Would you object to turning this into a "private" function `_is_bit_flag`? With the caveat that I haven't used bit flags extensively it seems like this is most useful internally.

mcara replied:
@mwcraig I was actually thinking the same thing a couple of hours earlier but I simply forgot about it. I also do not see any obvious use for this function (unless somebody would use it for development - not for direct use by users). It can always be made public later if there enough interest. It is more difficult to make it private later.

Probably I have developed this function to be used instead of https://github.com/MSeifert04/ccdproc/blob/81eea65e161168af2edd4d2acd8770623eacc53a/ccdproc/bitmask.py#L14 My original code did not have an extensive validation of inputs. I have a feeling a developer might benefit from this function if they will ever want to implement something like ["name maps"](https://github.com/MSeifert04/ccdproc/blob/81eea65e161168af2edd4d2acd8770623eacc53a/ccdproc/bitmask.py#L14) (also see "maps of flag names" in https://github.com/astropy/astropy/issues/4741#issue-143825905). So, if anyone developing something on top of `bitmask` would like to validate bit flags (provided as input or as values in a dict/map), this function might be useful. I will need to look at our own code here if I use it anywhere outside `bitmask`. If not - I am fine with making this function private (on the other hand making it visible will help avoiding people going through something like https://github.com/MSeifert04/ccdproc/blob/81eea65e161168af2edd4d2acd8770623eacc53a/ccdproc/bitmask.py#L14).

mcara replied:
Yep, I do not see this function being used anywhere else outside `bitmask`. @mwcraig Let me know if you do not think this function might be of use to _developers_. If you do - I'd say let's leave it public (???)

mwcraig replied:
I'd lean towards private since more dev types will probably dig into the source code anyway.

mcara commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+
+        .. note::
+
+            ``'None'``, ``'INDEF'``, and empty (or all white space) strings
```
@mwcraig I almost missed this issue... Support for string ``'None'`` and ``'INDEF'`` values is here because of backward compatibility that we need for HST software (and ``'INDEF'`` comes from IRAF). IMO we should not support this (I can try to take care of this this compatibility in our software outside of the `bitmask` module) in `astropy`. What is the policy here and what is your opinion? While this support is already implemented and so no extra effort is necessary, I am afraid to set a possibly bad precedent.

mwcraig replied:
Let's remove the support for 'None' and 'INDEF'; I don't know that there is a policy, but using `None` or an empty string seems clearer.

mcara commented:
@mwcraig In the light of changing the default of `good_mask_value` from `True` to `False`, would it also make more sense to rename `good_mask_value` to `bad_mask_value` (and change the default back to `True`)?

mwcraig commented:
> @mwcraig In the light of changing the default of good_mask_value from True to False, would it also make more sense to rename good_mask_value to bad_mask_value (and change the default back to True)?

I'm fine with either; maybe more important is to mention in the docstring that the default value provide the kind of mask numpy is looking for.

Once that and the `INDEF` change is done I think this is ready to go.

mcara commented:
@mwcraig I was so wrong in https://github.com/astropy/astropy/pull/7944#issuecomment-432401148 - I did not notice that I copied tests from `ccdproc` instead of more thorough tests that I later implemented in [`stsci.tools.tests.test_bitmask.py`](https://github.com/spacetelescope/stsci.tools/blob/4f9a01241a180d5849f02048d0fdc321de1c0b0b/lib/stsci/tools/tests/test_bitmask.py)

In my latest commit I did the following:

1. Added narrative help;
2. Added tests from `stsci.tools.tests.test_bitmask.py`;
3. Changed the default of `good_mask_value` from `True` to `False`.

I also changed my mind and decided to keep `'None'` and `'INDEF'`: I see no harm in keeping them and they are useful when processing either string user input or configuration files.

mcara commented:
@bsipocz Rebased

mwcraig commented:
Looks like one real failure remains at: https://travis-ci.org/astropy/astropy/jobs/446567968#L2151 (issue is a change in how numpy displays returned arrays (specifically whether the dtype is displayed).

Am going to do one more quick review, then will spend a minute tracking down the fix for the dtype issue from other recent-ish nddata PRs.

mwcraig commented:
Found an example of dealing with problematic doctest output: https://github.com/astropy/astropy/pull/6959/files#diff-ef1d13ad3ca4675e93a75fdf8461d387L1212

mcara commented:
@mwcraig Thanks for "continuous reviewing"! 😄 I'll fix the minor issues.

With regard to `nddata.utils`: that one seems to contain very specialized functions (I am not sure if they are even generally useful) such as `overlap_slices`, or even worse, `add_array` (when would anyone want to do that?) DQ->boolean mask is a much more common operation and, IMO, `bitmask` should stay by itself as it is a generally useful tool.

With regard to `dtype` in docs: I have no idea what is going on here. Locally I do not see `dtype` being printed at all. Do you get output with `dtype` when you run my example?

mwcraig commented on astropy/nddata/bitmask.py:
```diff
@@ -0,0 +1,411 @@
+        >>> bitmask.bitfield_to_boolean_mask(dqbits, ignore_flags=~6,
+        ...                                  good_mask_value=0, dtype=int)
+        array([[0, 0, 0, 1, 0, 0, 1, 0],
+               [1, 1, 0, 0, 0, 0, 1, 0]])
```
```suggestion
               [1, 1, 0, 0, 0, 0, 1, 0]]...)
```

I think this will fix the breaking tests

mcara commented:
> The prob is that whether dtype is displayed depends on the numpy version.

Should I then just change my examples to `print` arrays? It will complicate everything but it will not use the cryptic `...` from that workaround that you have linked to? When I read those examples, I wonder what was there (instead of `...`).

mcara commented:
Instead of `...` at the end of numpy arrays, what about using the suggestion from [Numpy 1.15.1 release notes](https://docs.scipy.org/doc/numpy-1.15.1/release.html#many-changes-to-array-printing-disableable-with-the-new-legacy-printing-mode)
```
try:
    np.set_printoptions(legacy='1.13')
except TypeError:
    pass
```
or, alternatively, something like this ("only" three lines):
```
from distutils.version import LooseVersion
if LooseVersion(np.version.short_version) >= LooseVersion('1.14'):
    np.set_printoptions(legacy='1.13')
```

mwcraig commented:
That might work....any idea whether that would affect any of the other tests? I guess I'm inclined to go the ugly `...` fix for today and defer to a later date implementing a more elegant fix package-wide.

mcara commented:
What "other" tests? (I am so lost in these huge logs)

mwcraig commented:
> What "other" tests? (I am so lost in these huge logs)

Sorry, meant the other tests in nddata. I guess it shouldn't since your tests are in a separate file. Go with whichever you prefer (the numpy-sanctioned way or `...`)

mwcraig commented:
@mcara do you have time to make the remaining changes today? If not, I can do them so we get this in for feature freeze.

mcara commented:
@mwcraig I am sorry, what other changes... And I am sorry if I look lost but I feel I am missing something. I just made a new commit that I think fixes all issues (I hope).

mcara commented:
@mwcraig https://github.com/astropy/astropy/pull/7944/commits/8e3179716c25aa1c58411e378fe4b0c3af3e62ef

mwcraig commented:
> @mwcraig I am sorry, what other changes... And I am sorry if I look lost but I feel I am missing something. I just made a new commit that I think fixes all issues (I hope).

Turns out there was a race condition....I was writing my comment at about the same time that you were pushing the commit 🤣

mcara commented:
@mwcraig Ah, I had a typo in the last commit: `opposited`. Fixed in the latest commit: `opposed`.

mcara commented:
Yes, we need a semaphore

mcara commented:
I was able to test (doctest) the `rst` locally. I hope `bitmask.py` will work in CircleCI

mwcraig commented:
Asking on slack about the test failures on travis, which look unrelated...

mcara commented:
Thank you: I am ready to give up this ... (:

mwcraig commented:
Feature freeze day often has special challenges; I'm sure we'll get this in even if we don't get the test passing today. I really appreciate this contribution -- it is important functionality. If there are any more issues I'll patch them up by adding to your PR (you shouldn't need to do anything to allow this to happen, it is the default github sets when you open a pull request)

mcara commented:
@mwcraig Thank you very much for your numerous reviews and taking care of this PR.

mcara commented:
@mwcraig I wonder whether I should rebase (again)?

mwcraig commented:
It can't hurt...

mcara commented:
Done

mwcraig commented:
I'm about to walk home, will dig into this some more in a couple of hours. No clue what is up with the travis failure...

mcara commented:
However, it just does not appear related to this PR. I see no connection.

mwcraig commented:
Agreed, it makes no sense at all...

pllim commented:
My hunch is that your `np.set_printoptions(legacy='1.13')` is messing with the rest of the doctest. Either:

* Set it back to non-legacy when you don't need it; OR
* Modify your doctest so you don't need that setting.

I have not seen this error in all the other PRs, so it must be related to your changes.

mcara commented:
@pllim Do you know if setting `np.set_printoptions(legacy='1.13')` in e.g.,  `bitmask.py` persists across all other packages and modules across `astropy`? This would be weird. For example, let's assume I remove `np.set_printoptions(legacy='1.13')` **AND** `import numpy as np`... Will my examples still work because (I am sure) somewhere before someone already imported `numpy` and so my example will run correctly (with a missing import)? That would be crazy!

mcara commented:
> Modify your doctest so you don't need that setting.

That's not really an option: it appears that the only way not to get `dtype` at the end of printed array is to have a mask of type `int`. It would be weird in a module that create "boolean masks" not to have at least a couple of examples with boolean type...

mcara commented:
@mwcraig If it is OK with you (we are already at 9th commit) I can try removing `np.set_printoptions(legacy='1.13')` and making all tests `int` just for the fun of it...

pllim commented:
> That would be crazy!

It is a crazy world! That is why we have astropy/pytest-doctestplus#15 open.

mwcraig commented:
I'm able to reproduce the issue on my computer; it is definitely related to the `np.set_printoptions(legacy='1.13')`; in the commit before that the errors don't happen, and in that commit running the tests for just the packages that fail on travis (e.g. units) then those tests pass. Just tried running with numpy 1.13 and all of the tests pass, so the print_options thing is the cause it looks like.

(Along the way I did figure out an unrelated problem that came up in a different context. Turns out using conda to remove numpy 1.15 (but not numpy-base which I didn't even know was a thing) leaves your numpy in a very confused state).

mwcraig commented:
I think using an ellipsis at the end should work, am trying it locally...

mcara commented:
OK. Then we can restore `np.set_printoptions(legacy=False)` at the end of the test.

pllim commented:
Re: `numpy-base` -- I know, right?! I wasted 30 mins trying to fix that mess on my machine just now.

mwcraig commented:
> OK. Then we can restore np.set_printoptions(legacy=False) at the end of the test.

That will leave the doctest examples a little convoluted, though...

mwcraig commented:
@pllim -- can we skip doctests on some numpy versions? That would in some ways be a cleaner solution.

mwcraig commented:
Actually, I may be wrong about the tests passing on numpy 1.13, forgot I had also made a source code change.

mcara commented:
I hate ellipses but I also hate alternatives. I think I will just switch to ellipses and be done with this. We can figure out something later OR `np v1.13` be dropped sooner than later

mcara commented:
> Turns out using conda to remove numpy 1.15 (but not numpy-base which I didn't even know was a thing) leaves your numpy in a very confused state)

Isn't conda such an amazing gift?

mwcraig commented:
> Isn't conda such an amazing gift?

I've gotten in the habit of almost always working in clean (i.e. single or short-term use) environments...in any event will ask about numpy-base because I don't really understand its purpose.

mwcraig commented:
the ellipses look like they will work

mcara commented:
@mwcraig I am testing it now in `np v1.15.2`. What I really, really, don't like about it is how this line will look https://github.com/astropy/astropy/pull/7944/files#diff-91a3a6f0f11094346c450f553e44cc89R110:


```
>>> bitmask.bitfield_to_boolean_mask(217, ignore_flags=80)
array(True...)
```
🤔 what `...` stands for...?

pllim commented:
> can we skip doctests on some numpy versions?

AFAIK no. But @drdavella would know better...

mwcraig commented:
> 🤔 what `...` stands for...?

Agreed, it is pretty terrible

pllim commented:
> can we skip doctests on some numpy versions?

*But* you can brute force it, by doing a numpy version check before running the code...?

```python
>>>if NUMPY_GT_1_3:
 ...     bitmask.something_something
```

mcara commented:
OK, tests pass in `1.15.2` using ellipses (they also were passing with `np.set_printoptions(legacy=False)`). What do I do? Is there a consensus?

Only not 

```
>>>if NUMPY_GT_1_3:
 ...     bitmask.something_something
```

I do not want to duplicate 20 tests just to have a version of array with `dtype=bool`.

Another alternative is to have all mask using `int`.

What do you think? We need to make a decision...

mcara commented:
Done

mwcraig commented:
🤞

mcara commented:
I am glad you cannot see the tear in my eye.

mwcraig commented:
:tada:

<img width="1002" alt="screenshot_10_26_18__9_46_pm" src="https://user-images.githubusercontent.com/1147167/47598887-b3baf380-d968-11e8-95ec-f79247c9fd6d.png">

mcara commented:
Thank you @mwcraig @pllim !

mwcraig commented:
We'll wait for things to completely finish before merging, but all signs are positive...

mwcraig commented:
(signing off for the night while things are working 🤣 )

bsipocz commented:
All green and approved, so I'm merging. 

Thank you @mcara  and @mwcraig!
